server/src/main/java/org/elasticsearch/index/engine/ThreadPoolMergeScheduler.java

```diff
@@ -0,0 +1,440 @@
+        // forced merges, as well as merges triggered when closing a shard, always run un-IO-throttled
+        boolean isAutoThrottle = mergeTrigger != MergeTrigger.CLOSING && merge.getStoreMergeInfo().mergeMaxNumSegments() == -1;
+        // IO throttling cannot be toggled for existing merge tasks, only new merge tasks pick up the updated IO throttling setting
+        return new MergeTask(
+            mergeSource,
+            merge,
+            isAutoThrottle && config.isAutoThrottle(),
+            "Lucene Merge Task #" + mergeTaskCount.incrementAndGet() + " for shard " + shardId
+        );
+    }
+
+    // synchronized so that {@code #currentlyRunningMergeTasks} and {@code #backloggedMergeTasks} are modified atomically
+    private synchronized boolean runNowOrBacklog(MergeTask mergeTask) {
+        assert mergeTask.isRunning() == false;
+        if (currentlyRunningMergeTasks.size() >= config.getMaxThreadCount()) {
+            backloggedMergeTasks.add(mergeTask);
+            return false;
+        } else {
+            boolean added = currentlyRunningMergeTasks.add(mergeTask);
+            assert added : "starting merge task [" + mergeTask + "] registered as already running";
+            return true;
+        }
+    }
+
+    private void mergeDone(MergeTask mergeTask) {
+        assert mergeTask.isRunning();
+        synchronized (this) {
+            boolean removed = currentlyRunningMergeTasks.remove(mergeTask);
+            assert removed : "completed merge task [" + mergeTask + "] not registered as running";
+            // when one merge is done, maybe a backlogged one can now execute
+            enqueueBacklogged();
+        }
+        mergeTaskDoneCount.incrementAndGet();
+        checkMergeTaskThrottling();
+    }
+
+    private synchronized void enqueueBacklogged() {
+        int maxBackloggedTasksToEnqueue = config.getMaxThreadCount() - currentlyRunningMergeTasks.size();
+        while (maxBackloggedTasksToEnqueue-- > 0) {
+            MergeTask backloggedMergeTask = backloggedMergeTasks.poll();
+            if (backloggedMergeTask == null) {
+                break;
+            }
+            threadPoolMergeQueue.enqueueMergeTask(backloggedMergeTask);
+        }
+    }
+
+    /**
+     * Does the actual merge, by calling {@link org.apache.lucene.index.MergeScheduler.MergeSource#merge}
+     */
+    protected void doMerge(MergeSource mergeSource, MergePolicy.OneMerge merge) throws IOException {
+        mergeSource.merge(merge);
+    }
+
+    @Override
+    public Directory wrapForMerge(MergePolicy.OneMerge merge, Directory in) {
+        // Return a wrapped Directory which has rate-limited output.
+        // Note: the rate limiter is only per thread. So, if there are multiple merge threads running
+        // and throttling is required, each thread will be throttled independently.
+        // The implication of this, is that the total IO rate could be higher than the target rate.
```

Comment:
I think that sounds fine, the target rate will auto-adjust to a good level per thread?

Also, this means the minimum target rate is per thread, which sort of also makes some sense in relation to ergonomics

Reply:
>I think that sounds fine, the target rate will auto-adjust to a good level per thread?

The target IO rate is adjusted based on the amount of "active" (running and queued) merge tasks, that also takes into account the maximum number of threads executing merges (`threadPool.info(ThreadPool.Names.MERGE).getMax()`).
The effective IO rate per ES node (per disk) is `threadPool.info(ThreadPool.Names.MERGE).getMax() * targetMBPerSec`.

Reply:
Here how I view it:

In principle, given a prototypical "merge load" the `targetMBPerSec` IO rate should stabilize to some lower value given higher values for `threadPool.info(ThreadPool.Names.MERGE).getMax()`.
The idea is to keep the merge threads always busy "in the background", meaning that they are allotted just enough IO rate in order to not fall behind, irrespective of `threadPool.info(ThreadPool.Names.MERGE).getMax()`, which is what you've suggested, I remember (It's probable that we can use the segment merge sizes to better inform the algorithm, but in principle it should work similarly).

Higher values for `threadPool.info(ThreadPool.Names.MERGE).getMax()` help when merges are not bound by IO rate, but instead need more CPU.

Reply:
Pushed a9f48235